app/src/main/java/com/tomaszpolanski/androidsandbox/ViewModelBase.kt

```diff
@@ -0,0 +1,25 @@
+package com.tomaszpolanski.androidsandbox
+
+import android.arch.lifecycle.ViewModel
+import android.os.Parcelable
+import io.reactivex.Flowable
+import io.reactivex.Single
+import io.reactivex.processors.PublishProcessor
+
+abstract class ViewModelBase<S : Parcelable, in E>(
+    private val schedulerProvider: SchedulerProvider
+                                                  ) : ViewModel() {
+
+    private val intentPublishSubject = PublishProcessor.create<Pair<E, S?>>()
+
+    fun getState(restored: S?, initial: S): Flowable<S> = intentPublishSubject
+        .flatMapSingle { (event, previous) ->
+            reduce(previous ?: restored ?: initial, event)
+        }
+        .observeOn(schedulerProvider.ui())
```

Comment:
IMO, state updates (the reduce function) should always happen on the UI thread. otherwise you'll have multiple threads touching the state.

Reply:
Reduce returns Singles and it can change the schedulers. I do not think we should limit that.
Are you worried about context switching/performance? If so, let's first measure the performance before we optimize it

Reply:
I'm not worried about performance, im worried about what happens when thread A and thread B both try to update the state at the same time :) Which will win? This function should ensure that it only ever happens on one thread.

IMO, reduce should be a pure function and not have any side effects.